queues/classic_queue.cc

```diff
@@ -66,7 +67,8 @@ template <typename T> class ts_queue {
       return false;
     Data = Buffer[NRel % Buffer.size()];
     NRel = (NRel + 1) % Buffer.size();
-    NCur -= 1;
+    --NCur;
+    std::clog << "-";
     Lk.unlock();
     CondProd.notify_one();
```

Comment:
This only works outside of mutex because we changed the same condition `CondProd` is waiting for. If we had no condition on the variable, it would miss signals sometimes :skull: . Same applies to `notify_all()`.